### PR TITLE
Private events app

### PR DESCRIPTION
**Pull request by notlfish**
Pull request to add a private events app.
The app should implement:
- User and Event models, with their corresponding information.
- Associations between users and events:
  - A user can host many events.
  - An event has a host user.
  - A user can attend many events.
  - An event can have many guests.
- CRUD actions for events.
- User creation action.
- Authentication and log in.
- Unit and feature tests using rspec and capybara.

### CODE REVIEW

**Review by Georjane — CHANGES_REQUESTED** [review]
### Hello, Team <STUDENTS>, ✋🏽 

Good job so far! You have a very professional Readme file, your PR is professional and you have a good mastery of git-flow. Your linters are all passing, I applaud that too.
However, we have a few more changes to make in order to meet all requirements for this project but you are almost there!

## Required Changes ♻️

- [x] After successful sign in, a user should be redirected to the Event's index page (set to `root_path` in the `Routes` file) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/login_requirement.\*\*png\*\*) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)) and not to create new events as it is in your app
- [x] The join event button throws an error
image
The event's show page displays the list of attendants but  we can't see this because of the error above

#### Controllers

##### EventsController

###### `create` action

* [x] Uses the **#build** [association refrence method](https://guides.rubyonrails.org/association_basics.html#detailed-association-reference), i.e. `current_user.events.build(event_params)` to create the new event with the user’s ID prepopulated.([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

###### `index` action

* [x] Lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] Lists upcoming events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### UsersController

###### `show` action

* [x] Lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] Lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] Lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] No notifications is been shown when a user that doesn't exist tries to sign in


Cheers and Happy coding! 💻 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 🤝 

_Code Reviewer
**[Jane](https://witahgeorjane.tech/)**_

File: app/models/event.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  has_many :event_attendances, foreign_key: 'attended_event_id', dependent: :destroy
+  has_many :attendees, through: :event_attendances
+
+  # def self.past_events_of(user)
+  #   user.attended_events.where(["date < ?", DateTime.now])
+  # end
+
+  # def self.future_events_of(user)
+  #   user.attended_events.where(["date > ?", DateTime.now ])
+  # end
```
Comment (SajjadAhmad14) [inline app/models/event.rb]: Please remove commented code from here.

File: app/views/devise/mailer/email_changed.html.erb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
+  <p>We're contacting you to notify you that your email is being changed to <%= @resource.unconfirmed_email %>.</p>
+<% else %>
```
Comment (SajjadAhmad14) [inline app/views/devise/mailer/email_changed.html.erb]: Please move this logic to app/helpers. [ OPTIONAL ]

File: app/views/devise/passwords/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+  <%= f.hidden_field :reset_password_token %>
+
+  <div class="field">
+    <%= f.label :password, "New password" %><br />
+    <% if @minimum_password_length %>
+      <em>(<%= @minimum_password_length %> characters minimum)</em><br />
+    <% end %>
```
Comment (SajjadAhmad14) [inline app/views/devise/passwords/edit.html.erb]: Please move this logic to app/helpers. [ OPTIONAL ]

File: app/views/devise/shared/_error_messages.html.erb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+                 resource: resource.class.model_name.human.downcase)
+       %>
+    </h2>
+    <ul>
+      <% resource.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+    </ul>
+  </div>
+<% end %>
```
Comment (SajjadAhmad14) [inline app/views/devise/shared/_error_messages.html.erb]: Please move this logic to app/helpers. [ OPTIONAL ]

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -1,16 +1,34 @@
...
+  <div class="navbar">
+    <ul class="nav-list">
+      <li><%=link_to "Home", root_path, class:"home"%></li>
+      <% if user_signed_in?  %>
+      <li><%= link_to 'Log Out', destroy_user_session_path, method: 'delete', class:"log" %></li>
+      <li><%= link_to current_user.name, user_path(current_user), class:"user" %></li>
+      <% else %>
+      <li><%= link_to 'Sign Up', new_user_registration_path, class:"log" %></li>
+      <li><%= link_to 'Log In', new_user_session_path, class:"log" %></li>
+      <% end %>
```
Comment (SajjadAhmad14) [inline app/views/layouts/application.html.erb]: Please move this logic to app/helpers. [ OPTIONAL ]

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, thank you for working on changes suggested by the previous reviewer :ok_hand:
## To Highlight 🎉
- [ ] Git flow is used properly.
- [ ] App is working perfectly.
## To Improve 🙏
- [x] Please remove commented code. It makes the overall codebase unclean.
- [ ] Please add the `npm install` or `yarn install` command in the Setup section and `rspec` to run tests in README.
- [x] Please see inline comments for reference.
# Optional
- [ ] Views should be completely separate from app logic. You can move all view-related logic to app/helpers.
### Happy Coding ☺️

## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

**Review by cyonii — APPROVED** [review]
## STATUS: APPROVED ✅

You have a really neat project 👏.

Congratulations 🎉, your project has been approved.
You have met the requirements for this project, you can now merge the PR.

> #### <br>Feel free to leave a comment or question if something is not clear. Slack: @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR)<br><br>

**Pull request merged**